arangod/Cluster/Utils/ShardID.cpp

```diff
@@ -0,0 +1,83 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2023-2023 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Michael Hackstein
+////////////////////////////////////////////////////////////////////////////////
+
+#include "ShardID.h"
+
+#include "Basics/Exceptions.h"
+#include "Basics/StringUtils.h"
+
+#include <absl/strings/str_cat.h>
+
+arangodb::ResultT<arangodb::ShardID> arangodb::ShardID::shardIdFromString(
+    std::string_view s) noexcept {
+  if (!s.starts_with('s')) {
+    return Result{TRI_ERROR_BAD_PARAMETER,
+                  "Expected ShardID to start with 's'"};
+  }
+  auto res = basics::StringUtils::try_uint64(s.data() + 1, s.length() - 1);
+  if (res.fail()) {
+    return std::move(res.result());
+  }
+  return ShardID{res.get()};
+}
+arangodb::ShardID arangodb::ShardID::invalidShard() noexcept { return ShardID{0}; }
+arangodb::ShardID::ShardID(std::string_view id) {
+  auto maybeShardID = shardIdFromString(id);
+  if (maybeShardID.fail()) {
+    THROW_ARANGO_EXCEPTION(maybeShardID.result());
+  }
+  *this = maybeShardID.get();
+}
+arangodb::ShardID::operator std::string() const {
+  return absl::StrCat("s", std::to_string(_id));
+}
+
+// NOTE: This is not noexcept because of shardIdFromString
+bool arangodb::ShardID::operator==(std::string_view other) const noexcept {
```

Comment:
Exactly, it is not noexcept because of the call to `shardIdFromString`. So we should remove the `noexcept` here.